File: support/handmade/.json/jsonDepot.cs
// 以下是模板驗證用程式碼

using System.Collections.Generic;

namespace SheeterJson {
    public partial class Depot {
        public Loader Loader { get; set; }
        public readonly RewardReader Reward = new RewardReader();
        private readonly List<Reader> Readers = new List<Reader>();

        public Depot() {
            Readers.Add(Reward);
        }

        public bool FromData() {
            if (Loader == null)
                return false;

            var result = true;

            foreach (var itor in Readers) {
                var filename = itor.FileName();
                var data = Loader.Load(filename);

                if (data == null || data.Length == 0)
                    continue;

                var message = itor.FromData(data);

                if (message.Length != 0) {
                    result = false;
                    Loader.Error(filename.File, message);
                }
            }

            return result;
        }

        public bool MergeData() {
            if (Loader == null)
                return false;

            var result = true;

            foreach (var itor in Readers) {
                var filename = itor.FileName();
                var data = Loader.Load(filename);

                if (data == null || data.Length == 0)
                    continue;

                var message = itor.MergeData(data);

                if (message.Length != 0) {
                    result = false;
                    Loader.Error(filename.File, message);
                }
            }

            return result;
        }

        public void Clear() {
            foreach (var itor in Readers) {
                itor.Clear();
            }
        }
    }

    public class FileName {
        public FileName(string name, string ext) {
            this.name = name;
            this.ext = ext;
        }

        public string Name {
            get {
                return name;
            }
        }

        public string Ext {
            get {
                return ext;
            }
        }

        public string File {
            get {
                return name + "." + ext;
            }
        }

        private readonly string name;
        private readonly string ext;
    }

    public interface Loader {
        public void Error(string name, string message);
        public string Load(FileName filename);
    }

    public interface Reader {
        public FileName FileName();
        public string FromData(string data);
        public string MergeData(string data);
        public void Clear();
    }
}

// 以下是為了通過編譯的程式碼, 不可使用

File: support/handmade/.json/jsonDepot.go
package _json

type Depot struct {
	Reward  RewardReader
	readers []ReaderInterface
}

func (this *Depot) FromData(load DepotLoad, error DepotError) bool {
	this.build()
	result := true

	for _, itor := range this.readers {
		data := load(itor.DataName(), itor.DataExt())

		if data == nil || len(data) == 0 {
			continue
		}

		if err := itor.FromData(data); err != nil {
			result = false
			error(itor.DataName(), err)
		}
	}

	return result
}

func (this *Depot) MergeData(load DepotLoad, error DepotError) bool {
	this.build()
	result := true

	for _, itor := range this.readers {
		data := load(itor.DataName(), itor.DataExt())

		if data == nil || len(data) == 0 {
			continue
		}

		if err := itor.MergeData(data); err != nil {
			result = false
			error(itor.DataName(), err)
		}
	}

	return result
}

func (this *Depot) build() {
	if len(this.readers) == 0 {
		this.readers = append(
			this.readers,
			&this.Reward,
		)
	}
}

type DepotError func(name string, err error)
type DepotLoad func(name, ext string) []byte

type ReaderInterface interface {
	DataName() string
	DataExt() string
	DataFile() string
	FromData(data []byte) error
	MergeData(data []byte) error
}
